docs/typescript.md

```diff
@@ -0,0 +1,220 @@
+---
+id: typescript
+title: Using TypeScript with React Native
+---
+
+[TypeScript][ts] is a language which extends JavaScript to add type definitions, much like [Flow][flow]. While React Native is built in Flow, it supports both TypeScript _and_ Flow by default.
+
+## Getting Started with TypeScript
+
+If you're starting a new project, there are a few different ways to get started. You can use the [default TypeScript template][ts-template]:
+
+```sh
+react-native init MyTSProject --template typescript
+```
+
+You can use [Expo][expo] which has two TypeScript templates:
+
+```sh
+npm install -g expo-cli
+expo init MyTSProject
+```
+
+Or you could use [Ignite][ignite], which also has a TypeScript template:
+
+```sh
+npm install -g ignite-cli
+ignite new MyTSProject
+```
+
+## Adding TypeScript to an Existing Project
+
+1. Add TypeScript and the React Native and Jest types to your project. (When you were using JavaScript these could have been added for you by your editor.)
+
+```sh
+yarn add typescript @types/jest @types/react @types/react-native @types/react-test-renderer
+# or for npm
+npm install --save-dev @types/jest @types/react @types/react-native @types/react-test-renderer
+```
+
+2. Add a TypeScript config file. Create a `tsconfig.json` in the root of your project:
+
+```json
+{
+  "compilerOptions": {
+    "allowJs": true,
+    "allowSyntheticDefaultImports": true,
+    "esModuleInterop": true,
+    "isolatedModules": true,
+    "jsx": "react",
+    "lib": ["es6"],
+    "moduleResolution": "node",
+    "noEmit": true,
+    "strict": true,
+    "target": "esnext"
+  },
+  "exclude": [
+    "node_modules",
+    "babel.config.js",
+    "metro.config.js",
+    "jest.config.js"
+  ]
+}
+```
+
+3. Create a `jest.config.js` file to configure Jest to use TypeScript
+
+```js
+module.exports = {
+  preset: 'react-native',
+  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx', 'json', 'node'],
+};
+```
+
+4. Rename a JavaScript file to be `*.tsx`
+
+5. Run `yarn tsc` to type-check your new TypeScript files.
+
+## How TypeScript and React Native works
+
+You should not rely on TypeScript to transform your `ts` and `tsx` files into JavaScript! Transforming your files to JavaScript works via the same [Babel infrastructure][Babel] as vanilla JavaScript. If you have existing TypeScript code, there are [one or two caveats][babel-7-caveats].
```

Comment:
Could we rephrased "Transforming your files to JavaScript works via the same Babel infrastructure as vanilla JavaScript" to be a bit clearer? For folks not familiar with how Babel works with vanilla JS (should we even still be calling it "vanilla JS?" Can we use a different phrase?)